### PR TITLE
Feat: Scheduled executions (execute later)

### PR DESCRIPTION
**Pull request by Meldiron**
## What does this PR do?

Adds parameter  to `createExecution` to allow execution in the future, similarly to how messages can also be scheduled for later.

## Test Plan

- [x] Implement tests

## Related PRs and Issues

x

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (Meldiron) [inline tests/e2e/Services/Functions/FunctionsCustomClientTest.php:259]: Let's sleep here for 1.5 mins, (if future time is 1min), and let's check it was completed after that.

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleFunctions.php:65]: In messaigng we do something a bit different:
https://github.com/appwrite/appwrite/blob/feat-scheduled-executions/src/Appwrite/Platform/Tasks/ScheduleMessages.php#L34-L36

Sould we do the same? I think this method runs every minute.. I am worried that with this logic, if you schedule function in 5 mins, it will run 5 times. Maybe.
Also scarred if worker is down, and it gets negative number here, god knows what happens.

I think it's the best bet to copy 99% of code from messaging - messaging scheduling works great.

File: app/controllers/api/functions.php
Comment (christyjacob4) [inline app/controllers/api/functions.php:1528]: Might be more relevant to use the BAD_REQUEST error here

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/ScheduleFunctions.php:71]: Where is $scheduleKey defined? I cant seem to find it

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (christyjacob4) [reply tests/e2e/Services/Functions/FunctionsCustomClientTest.php:259]: why 1.5 minutes? Cant we schedule something sooner? Like 15-20 seconds? Tests are already slow so lets try not to introduce more delays

Comment (Meldiron) [reply tests/e2e/Services/Functions/FunctionsCustomClientTest.php:259]: I think there is 10s sync interval, and 60s loop interval in scheduler.
Good catch, this makes function scheduled in 15 seconds from now only execute after a minute, I think. Let's check. Lets understand how quickly it can execute.

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleFunctions.php:71]: In the foreach above

File: app/controllers/api/functions.php
Comment (loks0n) [reply app/controllers/api/functions.php]: done

File: src/Appwrite/Event/Func.php
Comment (Meldiron) [inline src/Appwrite/Event/Func.php:226]: There is already $function, why cant we get function ID as `function->getId()`?

File: src/Appwrite/Platform/Workers/Functions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Workers/Functions.php:86]: Related to previous comment, we can `$function->getId()` I think

File: src/Appwrite/Platform/Workers/Functions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Workers/Functions.php:97]: We could do `getDocument` when we added `setFunctionId()` and do `setFunction` instead - preventing need for this

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:13]: Those numbers look very random. What about 5 and 5? 3 seems a bit too fast, considering we might have many of those worker, causing bunch of SQL queries just from here.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:28]: If I understand it right, we dont care about schedule cron here. Also I think it would never be valid cron in this worker - its datetime

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:40]: I think we don't need to do this in `go`. We do it for CRONs because we delay it some more with sleep. But here it's instant execution - we can do it without go.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:53]: setType, setMethod and setPath doesnt sound right. Method and path should be coming from the execution itself - this worker isn't setting it.
setType might in theory stay, but behaviour is closer to HTTP execution, compared to scheduled one.

File: src/Appwrite/Event/Func.php
Comment (loks0n) [reply src/Appwrite/Event/Func.php:226]: The `resource` of the scheduler is an `execution`, not a `function`. We aren't able to pass the whole `function` resource as we only have the linked `functionId` from the `execution`.

File: src/Appwrite/Platform/Workers/Functions.php
Comment (loks0n) [reply src/Appwrite/Platform/Workers/Functions.php:97]: We don't have `dbForProject` injected the scheduler, only `dbForConsole`.

File: src/Appwrite/Platform/Workers/Functions.php
Comment (loks0n) [reply src/Appwrite/Platform/Workers/Functions.php:86]: Please see above.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:13]: These numbers are copied from messaging scheduling, which works well. I don't think `5` and `3` will offer any difference?

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:28]: Yes - removed.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:40]: OK - removed.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:42]: OK - fixed.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:53]: Removed:
```
                    ->setMethod('POST')
                    ->setPath('/')
```

Do you have a suggestion for the alternative to `->setType('schedule')`?

File: app/controllers/api/functions.php
Comment (Meldiron) [inline app/controllers/api/functions.php:1608]: Let's add test that expects this exception

File: app/controllers/api/functions.php
Comment (Meldiron) [inline app/controllers/api/functions.php:1727]: Depending on what Eldad says, we might always want HTTP here.

File: app/controllers/api/functions.php
Comment (Meldiron) [inline app/controllers/api/functions.php:1775]: Compared to event triggered above, we are losing details here, like: body, headers, path, method, user.
JWT we shouldn't set here (it might expire), functions worker should generate it if we pass user.

I think we should add new parameter to `schedules` collection to store additional data like those, and just forward the mto function with execution.

Comment (Meldiron) [reply app/controllers/api/functions.php:1775]: Once fixed, lets update test to ensure this. For example the function could receive body and log it - when we do getExecution, we can check it's logs and see if it includes the word.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:45]: Let's add a comment here, that we only set ID because we don't have dbForProject at this point. Worker will later get the whole function.

Comment (Meldiron) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:45]: I would like to use dbForPRoject here, but this shouldn' be the reason to introduce it. When we have a comment, in future, when we add dbForProject for other reasons, we will improve this too.

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleFunctions.php:88]: Scheduled functions should still have those. That is coming from CRON executions, when they can't set those. Those needs to remain.

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:48]: As mention in API, here we need to also set stuff like body, headers, method, path, user jwt...
To get those data here, we could add attribute to schedules as temporary storage for those info (since we delete the document right after)

File: src/Appwrite/Event/Func.php
Comment (Meldiron) [inline src/Appwrite/Event/Func.php:15]: Do we need this change?

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (Meldiron) [inline tests/e2e/Services/Functions/FunctionsCustomClientTest.php:62]: Do we need this change?

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (Meldiron) [inline tests/e2e/Services/Functions/FunctionsCustomClientTest.php:259]: Not sure if 3rd parameter here is expected

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (Meldiron) [inline tests/e2e/Services/Functions/FunctionsCustomClientTest.php:281]: Not sure if 3rd parameter here is expected

File: app/controllers/api/functions.php
Comment (loks0n) [reply app/controllers/api/functions.php:1608]: done

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:45]: done

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (loks0n) [reply tests/e2e/Services/Functions/FunctionsCustomClientTest.php:281]: removed

File: tests/e2e/Services/Functions/FunctionsCustomClientTest.php
Comment (loks0n) [reply tests/e2e/Services/Functions/FunctionsCustomClientTest.php]: removed

File: src/Appwrite/Event/Func.php
Comment (loks0n) [reply src/Appwrite/Event/Func.php:15]: removed

Comment (loks0n) [reply src/Appwrite/Event/Func.php]: removed

File: src/Appwrite/Platform/Tasks/ScheduleFunctions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleFunctions.php]: reverted

File: app/controllers/api/functions.php
Comment (loks0n) [reply app/controllers/api/functions.php]: Added `metadata` column to hold this information in the `schedules` table

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:54]: Went with schedule for now

File: src/Appwrite/Platform/Tasks/ScheduleExecutions.php
Comment (Meldiron) [inline src/Appwrite/Platform/Tasks/ScheduleExecutions.php:31]: deleteDocument here too

Comment (loks0n) [reply src/Appwrite/Platform/Tasks/ScheduleExecutions.php:31]: done

File: app/controllers/api/functions.php
Comment (eldadfux) [inline app/controllers/api/functions.php:1595]: Not sure about this name, its in a past tense, but part of an action we're making now

**Review by eldadfux — CHANGES_REQUESTED** [review]

File: app/controllers/api/functions.php
Comment (Meldiron) [reply app/controllers/api/functions.php:1595]: Made a task to discuss after QA is out

File: app/config/collections.php
Comment (Meldiron) [reply app/config/collections.php:4563]: We need to add a column, but it doesn't require migration - patch script is enough. Like we did for some billing columns and tables.

File: app/controllers/api/functions.php
Comment (Meldiron) [reply app/controllers/api/functions.php:1785]: In another PR, for 1.6, we are making it so logging=false still creates document. We can ignore this bug, as it wont be there once all PRs are merged.

**Review by lohanidamodar — APPROVED** [review]
Approving this for Q/A release, confirming that the naming changes required will be taken care of during QA week.

**Pull request merged**
